lib/reader.ml

```diff
@@ -347,7 +347,9 @@ let parse_client_extension raw =
     match int_to_extension_type etype with
     | Some SERVER_NAME ->
        (match parse_hostnames buf with
-        | [name] -> `Hostname name
+        | [name] ->
+          let name = Domain_name.of_string_exn name |> Domain_name.host_exn in
```

Comment:
It may be better to use one of the `raise_*` functions here in the event of a bad domain name, instead of raising an Invalid_argument (which is what Domain_name.of_string_exn will do if there's a malformed domain passed to it). That'll make it easier for users of the library to reliably catch parsing exceptions.

Reply:
@avsm good point. The exposed API (reader.mli) includes `parse_handshake` -- which does a global `catch` (defined in reader.ml - translating some exceptions into `(_, err) result`). Certainly the treatment of `Invalid_argument` could be improved, or the failing `Domain_name.of_string_exn` should be translated to an informational error message. Certainly over the last 8 years the OCaml `result` ecosystem improved a lot, eventually we can even use some Stdlib operators.

Reply:
Thanks for the feedback @avsm ! I've cleaned up the exception handling here.